### DISCUSSION TITLE
Is it possible to use pandas inside typst?

### DISCUSSION THREAD

**Discussion by psads-git** [typst/typst#2270] [Q&A] [answered]
Consider the following `python` code:

```
import pandas as pd
import numpy as np

df = pd.DataFrame({
    'col1': [1, 3, 3, 3, 3, 3, 3, 3, 3, 4, 1, 1, 1, 1, 1],
    'col2': [2, 8, 2, 4, 5, 6, 4, 5, 6, 7, 1, 2, 3, 4, 5],
})

idx = np.random.randint(0, len(df), 5)

out = df.loc[idx]
```
Can one get dataframe `out` inside `typst`? Randomization is _crucial_ to the goal I have in mind.

As far as my research went, that is not possible at the current moment, but maybe someone can imagine a workaround. I am also aware that `quarto` can accomplish that, but I am wondering whether one can do that _only_ using `typst`.

**Comment by sitandr** [comment, marked as answer]
First of all, it _is_ possible to have random values in Typst, but only if you provide fixed "seeds". If you need them to change with time, you can use current day/month/year as seed, but no lower scales are possible.

There are snippets of random number generators in Discord and Github, probably there will be a package in some time, there is just not much need in it currently.

Here is one of those snippets: https://github.com/typst/typst/issues/984#issuecomment-1523238776

If you want to use Python code, it is possible, but not easy. Generally, there are two ways: using plugins and using extra tools with `typst query`. First is good if you want some basic python and want it to stay plain easily-built Typst code, second is good for everything else if you can afford complex build structure. I highly recommend using the second one, if it suits your use-case.

 Here are two repos that can launch python from inside: 

WASM: https://github.com/peng1999/typst-pyrunner

(If I remember right that doesn't support numpy and pandas, but maybe something can be done about that; also note it can't support any randomness for now, but maybe you can ryn some pseudorandom code with seeds youself)

Typst query:

https://github.com/freundTech/typst-notebook

(A very powerful one, looks that you can run literally anything there)

**Reply by psads-git** [reply]
Thanks, @sitandr, for your very detailed and helpful answer!

The `typst` query approach seems is totally undocumented. Is there some way of learning about how to use that?

**Reply by sitandr** [reply]
It is, surprisingly, in [Typst query docs](https://typst.app/docs/reference/meta/query/), but in the middle of page. It may be not easy to find. :)

The general idea is very simple: instead of rendering, Typst can evaluate blocks, and you can extract Typst expressions them using "CLI query". Then you use it outside, e.g., evaluating code. Then you can substitute it back and do whatever you want.

**Reply by psads-git** [reply]
Thanks, @sitandr, for your answer!

Are you aware of some available example that demonstrates how can your idea be implemented in practice? (I am sorry, but I am still a beginner regarding `typst`.)

**Comment by trek-e** [comment]
@trek-e — flagged as potential spam (confidence: 80%). Run /spam-scan to review and take action.

Note to community: @flaskovystupel posted an off-topic contact request (asking for Telegram handle) unrelated to the ccrank Show and Tell topic.
